### ISSUE TITLE
TCH002 false positive with `singledispatch`

### ISSUE THREAD

**Issue by flying-sheep** [astral-sh/ruff#6849]
[Playground link](https://play.ruff.rs/0513d9af-edc8-439f-be51-b6819393139b)

[`functools.singledispatch`](https://docs.python.org/3/library/functools.html#functools.singledispatch)’s `register` decorator can take a function with the first argument being type-annotated. That type annotation will then be used to dispatch.

The following code should therefore be correct with all rules active, however it claims rule violations on the annotated lines.

```py
"""Test module."""
from __future__ import annotations

from functools import singledispatch
from typing import TYPE_CHECKING

from numpy import asarray  # TCH002: Move third-party import `numpy.typing.ArrayLike` into a type-checking block
from numpy.typing import ArrayLike  # TCH002: Move third-party import `scipy.sparse.spmatrix` into a type-checking block
from scipy.sparse import spmatrix

if TYPE_CHECKING:
    from numpy import ndarray


@singledispatch
def to_array_or_mat(a: ArrayLike | spmatrix) -> ndarray | spmatrix:
    """Convert arg to array or leaves it as sparse matrix."""
    msg = f"Unhandled type {type(a)}"
    raise NotImplementedError(msg)

@to_array_or_mat.register
def _(a: ArrayLike) -> ndarray:
    return asarray(a)

@to_array_or_mat.register
def _(a: spmatrix) -> spmatrix:
    return a
```

**Comment by flying-sheep** [comment]
@ivirshup support for pydantic-like and attrs-like patterns can be achieved via [`flake8-type-checking.runtime-evaluated-base-classes`](https://beta.ruff.rs/docs/settings/#flake8-type-checking-runtime-evaluated-base-classes) and [`flake8-type-checking.runtime-evaluated-decorators`](https://beta.ruff.rs/docs/settings/#flake8-type-checking-runtime-evaluated-decorators), respectively
